Parser support for multi-table updates

bbeaudreault opened the pull request:
Let's see what travis has to say. Parser and planbuilder tests work locally.

I'll take a look at multi-table delete as well

bbeaudreault commented:
Heh I was tricked I guess by `go test` only doing a partial compile?  Will fix the compile issues.

michael-berlin commented:
`go test` within a package compiles only all dependencies :)

If you want to compile everything, I recommend to run `make build` in the root directory.

bbeaudreault commented:
I'm going to add tests for the PassDML stuff in tabletserver dml.go.  First I'm adding multi-delete parser support.

bbeaudreault commented:
I've added parser support and tests for multi-table delete

bbeaudreault commented:
Added `PASS_DML` tests.

@sougou this is ready for review. If I get a chance and am able to figure it out, I'll add vtgate planbuilder support. But could do that in a separate PR. If you have any guidance, it'd be appreciated.

sougou commented on go/vt/sqlparser/sql.y:
```diff
@@ -317,15 +319,33 @@ insert_or_replace:
-    $$ = &Delete{Comments: Comments($2), Table: $4, Where: NewWhere(WhereStr, $5), OrderBy: $6, Limit: $7}
+    $$ = &Delete{Comments: Comments($2), TableExprs:  TableExprs{&AliasedTableExpr{Expr:$4}}, Where: NewWhere(WhereStr, $5), OrderBy: $6, Limit: $7}
+  }
+| DELETE comment_opt table_name_list from_or_using table_references where_expression_opt
```
This doesn't look like [valid syntax](https://dev.mysql.com/doc/refman/5.7/en/delete.html). Did you have something else in mind?

bbeaudreault replied:
Which part is invalid? See the multi table syntax section there

sougou replied:
Sorry. Missed that part. It's weird syntax. I presume the `from` clause is used to identify rows, but the table list tells you the ones from which the rows have to be dropped.

If this is the case, the table list has to be a subset of what's in the `from` clause. Then the same processTableExprs on the `from` clause should work.

sougou commented on go/vt/vtgate/planbuilder/dml.go:
```diff
@@ -39,7 +39,17 @@ func buildUpdatePlan(upd *sqlparser.Update, vschema VSchema) (*engine.Route, err
+	updateTable, ok := aliased.Expr.(sqlparser.TableName)
+	if !ok {
+		return nil, errors.New("unsupported: complex table reference in update")
+	}
```
This section will nullify the feature because it will block all multi-table updates. Instead, you could call processTableExprs, which will return a builder. Then you can verify it with:
```
if rb, ok := bldr.(*route); !ok || rb.ERoute.Keyspace.Sharded {
  return nil, errors.New("unsupported: update statement spans multiple shards")
}
```
which will allow all single shard updates to pass through.

Same should work for deletes.

You'll need to add some tests in dml_cases.txt for the new capabilities.

bbeaudreault replied:
Yea I did this just to protect us before we added planbuilder support. But with your suggestion I'll work on that.

sougou replied:
I'm actually thinking we could be more permissive here. For example, we can convert `SelectEqualUnique` to `UpdateEqual`. So, instead of checking if keyspace is sharded, we can check route opcodes.
And by introducing a few more opcodes, we could perform multi-shard DMLs (in the future).

bbeaudreault replied:
Does this work on joins? We are type asserting to type `*route` but I believe a query like `update foo from foo join bar on foo.id = bar.id` would be of type `join`.

Also what should I do with the `bldr` after I assert this. I believe the next step would be to Push things onto it, but maybe we don't need that for update/delete?

bbeaudreault replied:
Ok, I see/remember that for unsharded keyspaces and where possible we will merge the joins into the existing route. Those are the only ones we want to target, so we can keep the `*route` assertion.

Still curious what to do with the bldr after checking for whether it's sharded?

bbeaudreault replied:
Ok I added an initial implementation. Will add some tests soon. Let me know what you think

bbeaudreault commented:
Working on fixing the tests, but also seeing some complications with regard to Route.Table, which is not set by processTableExprs but is required for DMLs

bbeaudreault commented:
What I've pushed still doesn't work, but I added pushing of filters, order by, limit, like we do in Select case.  I think we don't need `getDMLRouting` since we're building the full plan now.  Still trying to figure out what to do about Tables. Let me know if this is going in the wrong direction

bbeaudreault commented:
Tried using the symtab to handle generating the delete subquery.  I think I'm going to need to call Wireup, based on the results I'm seeing in the test output. But that is pretty specific to Select right now.

bbeaudreault commented:
@sougou per offline discussion, i dropped some of the past commits and went for the simpler approach for now.  Added tests, etc as well.  Ready for review. Let me know what you think

sougou commented on go/vt/vtgate/planbuilder/dml.go:
```diff
@@ -87,30 +106,51 @@ func isIndexChanging(setClauses sqlparser.UpdateExprs, colVindexes []*vindexes.C
+		er.Opcode = engine.DeleteUnsharded
+		return er, nil
+	}
+	updateTable, ok := aliased.Expr.(sqlparser.TableName)
```
We could follow the same logic as update here. It's also beneficial to keep the two flows similar (for readability). Instead, we can make some tweaks below:
* Everything up to the point where we check for unsharded keyspace remains similar to update.
* If the keyspace is sharded,  then we check `del.Targets != nil || er.Table == nil`. In other words, a unique table was not identified. So, we return an error.

I think this approach will work. Let me know what you think.

bbeaudreault commented on go/vt/vtgate/planbuilder/dml.go:
```diff
@@ -36,34 +36,53 @@ func dmlFormatter(buf *sqlparser.TrackedBuffer, node sqlparser.SQLNode) {
+		er.Opcode = engine.UpdateUnsharded
+		return er, nil
+	}
+	updateTable, ok := aliased.Expr.(sqlparser.TableName)
```
`Table` is used in `getDMLRouting`. So I pulled the table from the route's symtab.

sougou commented on go/vt/vtgate/planbuilder/dml.go:
```diff
@@ -36,34 +36,45 @@ func dmlFormatter(buf *sqlparser.TrackedBuffer, node sqlparser.SQLNode) {
+	for t := range rb.Symtab().tables {
+		tableName = t
+	}
+	er.Table, err = vschema.Find(tableName)
```
This sucks :). We can (later) fix this by adding a vindexTable field to the table in symtab.
